### PR TITLE
[mlir] Fix region simplification bug when later blocks use prior block argument values

### LINKED ISSUE #94520: [mlir] Incorrect region simplification in `applyPatternsAndFoldGreedily`
## Summary

Region simplification appears to be performing incorrect simplification under some nested loop cases, causing erroneous code generation. I have seen this incorrect behavior in 16.0.6 and 19.0.0 (on Ubuntu 22.04, x86_64). I have not checked any other versions.

## Details

The following MLIR causes incorrect region simplification in any pass that calls `applyPatternsAndFoldGreedily()` with `config.enableRegionSimplification` set to true (which is the default parameter value).

**nested_loop_bug.mlir**:
```mlir
module attributes {llvm.data_layout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128", llvm.triple = "x86_64-unknown-linux-gnu"} {
  func.func @nested_loop() -> i32 attributes {no_this} {
    %c3_i64 = arith.constant 3 : i64
    %c2_i64 = arith.constant 2 : i64
    %c0_i64 = arith.constant 0 : i64
    %c1_i64 = arith.constant 1 : i64
    %c1_i32 = arith.constant 1 : i32
    %c0_i32 = arith.constant 0 : i32
    cf.br ^bb1(%c0_i32, %c0_i64 : i32, i64)
  ^bb1(%0: i32, %1: i64):  // 2 preds: ^bb0, ^bb8
    %2 = arith.cmpi ult, %1, %c2_i64 : i64
    cf.cond_br %2, ^bb2(%0, %1 : i32, i64), ^bb9(%0, %1 : i32, i64)
  ^bb2(%3: i32, %4: i64):  // pred: ^bb1
    %5 = arith.addi %4, %c1_i64 : i64
    cf.br ^bb3(%3, %5 : i32, i64)
  ^bb3(%6: i32, %7: i64):  // 2 preds: ^bb2, ^bb5
    %8 = arith.cmpi ult, %7, %c3_i64 : i64
    cf.cond_br %8, ^bb4(%6, %7 : i32, i64), ^bb6(%6, %7 : i32, i64)
  ^bb4(%9: i32, %10: i64):  // pred: ^bb3
    %11 = arith.addi %9, %c1_i32 : i32
    cf.br ^bb5(%11, %10 : i32, i64)
  ^bb5(%12: i32, %13: i64):  // pred: ^bb4
    %14 = arith.addi %13, %c1_i64 : i64
    cf.br ^bb3(%12, %14 : i32, i64)
  ^bb6(%15: i32, %16: i64):  // pred: ^bb3
    cf.br ^bb7
  ^bb7:  // pred: ^bb6
    cf.br ^bb8(%15, %4 : i32, i64)
  ^bb8(%17: i32, %18: i64):  // pred: ^bb7
    %19 = arith.addi %18, %c1_i64 : i64
    cf.br ^bb1(%17, %19 : i32, i64)
  ^bb9(%20: i32, %21: i64):  // pred: ^bb1
    cf.br ^bb10
  ^bb10:  // pred: ^bb9
    return %20 : i32
  }
}
```

The above MLIR roughly corresponds to this C++ code (after running through some of our custom passes which are not relevant for the purposes of this issue):
```cpp
int nested_loop() {
  int nEle = 0;
  for (std::size_t r = 0; r < 2; r++)          // r = 0,1
    for (std::size_t s = r + 1; s < 3; s++)    // s = 1,2 then 2
      nEle += 1;                               // should be incremented 3 times
  return nEle;
}
```

The MLIR code can be executed without any passes using small test driver program, and it produces the correct result (3).

**nested_loop_driver.c**:
```cpp
#include <stdio.h>
int nested_loop();
int main(int argc, char *argv[]) {
  printf("nested_loop returns: %d\n", nested_loop());
  return 0;
}
```

### Results when using no additional passes
```bash
$ cat nested_loop_bug.mlir | mlir-opt-19 --convert-func-to-llvm --convert-cf-to-llvm | mlir-translate-19 --mlir-to-llvmir | llc-19 --relocation-model=pic --filetype=obj -o nested_loop_good.o
$ clang-19 nested_loop_driver.c nested_loop_good.o -o nested_loop_good.exe
$ ./nested_loop_good.exe
nested_loop returns: 3
```

In order to run `applyPatternsAndFoldGreedily` from the command line, we can simply use `--test-math-algebraic-simplification`. The problem is not unique to that pass, nor even really related to that pass because this MLIR will not match any patterns used by that pass. I'm merely using it to show how `applyPatternsAndFoldGreedily` is behaving.

### Results when using `applyPatternsAndFoldGreedily`
```bash
$ cat nested_loop_bug.mlir | mlir-opt-19 --test-math-algebraic-simplification --convert-func-to-llvm --convert-cf-to-llvm | mlir-translate-19 --mlir-to-llvmir | llc-19 --relocation-model=pic --filetype=obj -o nested_loop_bad.o
$ clang-19 nested_loop_driver.c nested_loop_bad.o -o nested_loop_bad.exe
$ ./nested_loop_bad.exe
nested_loop returns: 2
```

**The fact that is now returning 2 is incorrect operation.**

The following MLIR snippet shows the output of applyPatternsAndFoldGreedily (via --test-math-algebraic-simplification). 
```mlir
$ cat nested_loop_bug.mlir | mlir-opt-19 --test-math-algebraic-simplification
module attributes {llvm.data_layout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128", llvm.triple = "x86_64-unknown-linux-gnu"} {
  func.func @nested_loop() -> i32 attributes {no_this} {
    %c3_i64 = arith.constant 3 : i64
    %c2_i64 = arith.constant 2 : i64
    %c0_i64 = arith.constant 0 : i64
    %c1_i64 = arith.constant 1 : i64
    %c1_i32 = arith.constant 1 : i32
    %c0_i32 = arith.constant 0 : i32
    cf.br ^bb1(%c0_i32, %c0_i64 : i32, i64)
  ^bb1(%0: i32, %1: i64):  // 2 preds: ^bb0, ^bb7
    %2 = arith.cmpi ult, %1, %c2_i64 : i64
    cf.cond_br %2, ^bb2(%0, %1 : i32, i64), ^bb8(%0 : i32)
  ^bb2(%3: i32, %4: i64):  // 2 preds: ^bb1, ^bb4
    %5 = arith.addi %4, %c1_i64 : i64
    cf.br ^bb3(%3, %5 : i32, i64)
  ^bb3(%6: i32, %7: i64):  // pred: ^bb2
    %8 = arith.cmpi ult, %7, %c3_i64 : i64
    cf.cond_br %8, ^bb4(%6, %7 : i32, i64), ^bb5(%6 : i32)
  ^bb4(%9: i32, %10: i64):  // pred: ^bb3
    %11 = arith.addi %9, %c1_i32 : i32
    cf.br ^bb2(%11, %10 : i32, i64)
  ^bb5(%12: i32):  // pred: ^bb3
    cf.br ^bb6
  ^bb6:  // pred: ^bb5
    cf.br ^bb7(%12, %4 : i32, i64)
  ^bb7(%13: i32, %14: i64):  // pred: ^bb6
    %15 = arith.addi %14, %c1_i64 : i64
    cf.br ^bb1(%13, %15 : i32, i64)
  ^bb8(%16: i32):  // pred: ^bb1
    cf.br ^bb9
  ^bb9:  // pred: ^bb8
    return %16 : i32
  }
}
```

**Diff** (original on the left, applyPatternsAndFoldGreedily on the right)
image

It appears that `applyPatternsAndFoldGreedily` incorrectly merged `^bb5` and `^bb6` and correspondingly eliminated the `%14 = arith.addi %13, %c1_i64 : i64`. As demonstrated with the test program, this results in incorrect behavior.

### PR DESCRIPTION
**Pull request by bmhowe23**
This fixes #94520 by ensuring that any if any block arguments are being used outside of the original block that the block is not considered a candidate for merging.

More details: the root cause of the issue described in #94520 was that `^bb2` and `^bb5` were being merged despite `%4` (an argument to `^bb2`) was being used later in `^bb7`. When the block merge occurred, that unintentionally changed the value of `%4` for all downstream code. This change prevents that from happening.

### CODE REVIEW

**Comment by schweitzpgi** [comment]
Looks good to me.

**Comment by bmhowe23** [comment]
Pinging @bondhugula, @joker-eph, and @matthias-springer for review.

**Comment by bmhowe23** [comment]
Just FYI, this was tested locally by doing
```bash
$ cmake -S llvm -B build -G Ninja -DLLVM_ENABLE_PROJECTS="clang;lld;mlir" -DCMAKE_BUILD_TYPE=Release -DLLVM_BUILD_TESTS=ON -DMLIR_INCLUDE_INTEGRATION_TESTS=ON
$ cd build
$ ninja
$ ninja check-mlir
```
I know the last item exercised the new test. Without the change, it fails, and with the change, it passes.

**Comment by bmhowe23** [comment]
Ping. This is my first PR to this repo, so please let me know if I'm missing something about the process here or if I should rework any of this? Thanks!

**Comment by joker-eph** [comment]
Sorry, this just fell through the review crack. Feel free to ping every week if a PR does not get reviewers attention.

File: mlir/test/Integration/Dialect/ControlFlow/test-region-simplification.mlir [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+  cf.br ^bb1(%17, %19 : i32, i64)
+^bb9(%20: i32, %21: i64):  // pred: ^bb1
+  cf.br ^bb10
+^bb10:  // pred: ^bb9
+  return %20 : i32
+}
+
+// If region simplification behaves correctly (by NOT merging ^bb2 and ^bb5),
+// this will be 3.
+// CHECK: 3
```
Comment (joker-eph) [inline mlir/test/Integration/Dialect/ControlFlow/test-region-simplification.mlir:49]: Can you please replace the integration test by a unit-test: it should go in `mlir/test/Transforms/canonicalize-block-merge.mlir`. Thanks!

Comment (bmhowe23) [reply mlir/test/Integration/Dialect/ControlFlow/test-region-simplification.mlir:49]: OK, I can do that.

The command at the top of that file (`mlir-opt -allow-unregistered-dialect %s -pass-pipeline='builtin.module(func.func(canonicalize{region-simplify=aggressive}))'`) won't exercise the new code because if one runs canonicalization before region simplification, the problem is averted. The integration test that I ran intentionally added `enable-patterns=AnyPattern` to skip the typical canonicalization processing that masks the problem.

(I realize this is a bit convoluted, but the `mlir-opt` command line that I'm using is essentially trying to mimic what happens when MLIR code calls `applyPatternsAndFoldGreedily()`, hence the need to skip the regular canonicalizaiton.)

So I will need to add an additional `mlir-opt` command to the top of that file with a custom `FileCheck --check-prefix=PATTERNS`. Does that sound reasonable to you?

Also, I think the unit test would need to remain pretty much as complicated as the IR for the integration test in order to exercise this bugfix. Is it ok to simply copy the IR from the `test-region-simplification.mlir` and put it into the new unit test file? Or were you wanting a simpler test case? Any attempts to simplify it to a smaller number of blocks leads has thus far led to `error: operand #1 does not dominate this use`.

Comment (joker-eph) [reply mlir/test/Integration/Dialect/ControlFlow/test-region-simplification.mlir:49]: > So I will need to add an additional mlir-opt command to the top of that file with a custom FileCheck --check-prefix=PATTERNS. Does that sound reasonable to you?

Probably worth just adding a new file next to this one.

Test case should be the simplest, it's not clear to me about the source of the `error: operand #1 does not dominate this use`?  Is this the verifier on the input file or after the pass runs? If the input then that seems like an issue in the way you reduce it isn't it?

Comment (bmhowe23) [reply mlir/test/Integration/Dialect/ControlFlow/test-region-simplification.mlir:49]: Yes, I've discussed this a bit with @schweitzpgi. For a fully correct IR that truly demonstrates the bug, I believe it will require 6 basic blocks.
image

In this hypothetical case, bb2 and bb4 are identical and could be merged *except* if bb5 uses a block argument from bb2.

The scenario is fairly complicated, so that's why I originally opted for an integration test, because it definitively shows that the problem exists (before the bugfix).

Every attempt I've made to simplify it down to something smaller causes one of two things to happen: 1) the input validation fails because I've simplified it into an invalid test case, or 2) the other IR simplifications on the *output* makes it difficult to discern whether the bug has truly been resolved.

For example, regarding the first path (where input validation fails because I've simplified it into an invalid test case), if I construct a test case like this:
```mlir
func.func @nomerge2(%arg0: i32, %i: i32) -> i32 {
  %c1_i32 = arith.constant 1 : i32
  %icmp = arith.cmpi slt, %i, %arg0 : i32
  cf.cond_br %icmp, ^bb2(%arg0 : i32), ^bb3(%c1_i32 : i32)

^bb2(%bbarg2 : i32):
  %ip1 = arith.addi %bbarg2, %c1_i32 : i32
  cf.br ^bb4(%ip1 : i32)

^bb3(%bbarg3 : i32):
  %jp1 = arith.addi %bbarg3, %c1_i32 : i32
  cf.br ^bb4(%jp1 : i32)

^bb4(%j: i32):
  %jcmp = arith.cmpi slt, %j, %bbarg2 : i32
  return %bbarg2 : i32
}
```
Then the compiler correctly reports this error message and will not run the pass:
```
error: operand #1 does not dominate this use
  %jcmp = arith.cmpi slt, %j, %bbarg2 : i32
```

So if we have to construct valid IR for the input test case, then I think we will be left with a fairly complicated (and therefore difficult to interpret) test. So the integration test was an attempt to mitigate that by simply having a single numerical output tell you whether the bug has been fixed or not.

Comment (bmhowe23) [reply mlir/test/Integration/Dialect/ControlFlow/test-region-simplification.mlir:49]: FWIW, here is the smallest example I can construct that shows the problem in `main` and demonstrates that it is fixed with the changes in this PR:
```mlir
func.func @nested_loop() -> i32 {
  %c0_i32 = arith.constant 0 : i32
  %c1_i32 = arith.constant 1 : i32
  %c2_i32 = arith.constant 2 : i32
  %c3_i32 = arith.constant 3 : i32
  cf.br ^bb1(%c0_i32, %c0_i32 : i32, i32)
^bb1(%0: i32, %1: i32):
  %2 = arith.cmpi ult, %1, %c2_i32 : i32
  cf.cond_br %2, ^bb2(%0, %1 : i32, i32), ^bb7(%0 : i32)
^bb2(%3: i32, %4: i32):
  %5 = arith.addi %4, %c1_i32 : i32
  cf.br ^bb3(%3, %5 : i32, i32)
^bb3(%6: i32, %7: i32):
  %8 = arith.cmpi ult, %7, %c3_i32 : i32
  cf.cond_br %8, ^bb4(%6, %7 : i32, i32), ^bb6(%6, %4 : i32, i32)
^bb4(%9: i32, %10: i32):
  %11 = arith.addi %9, %c1_i32 : i32
  cf.br ^bb5(%11, %10 : i32, i32)
^bb5(%12: i32, %13: i32):
  %14 = arith.addi %13, %c1_i32 : i32
  cf.br ^bb3(%12, %14 : i32, i32)
^bb6(%15: i32, %16: i32):
  %17 = arith.addi %16, %c1_i32 : i32
  cf.br ^bb1(%15, %17 : i32, i32)
^bb7(%18: i32):
  return %18 : i32
}
```
If you place the above into `test.mlir` and run with the following command, the latest on `main` will return `2` while the fixed version returns `3`.
```bash
$ build/bin/mlir-opt --canonicalize='enable-patterns=AnyPattern region-simplify=aggressive' --convert-func-to-llvm --convert-cf-to-llvm test.mlir | build/bin/mlir-cpu-runner -e nested_loop --entry-point-result=i32
```

In this example, the original problem reported by the original issue occurs when `^bb2` and `^bb5` are merged.

Comment (bmhowe23) [reply mlir/test/Integration/Dialect/ControlFlow/test-region-simplification.mlir:49]: If I were to turn this into a unit test, the output (*) of the above test is difficult to prove that the problem is fixed, which I fear will make a unit test difficult to maintain going forward.

(*) Here is the output of the aforementioned `test.mlir` after running through the pass, and `test.mlir` is the simplest test case that I can make.
```mlir
$ mlir-opt --canonicalize='enable-patterns=AnyPattern region-simplify=aggressive' test.mlir
module {
  func.func @nested_loop() -> i32 {
    %c0_i32 = arith.constant 0 : i32
    %c1_i32 = arith.constant 1 : i32
    %c2_i32 = arith.constant 2 : i32
    %c3_i32 = arith.constant 3 : i32
    cf.br ^bb1(%c0_i32, %c0_i32 : i32, i32)
  ^bb1(%0: i32, %1: i32):  // 2 preds: ^bb0, ^bb5
    %2 = arith.cmpi ult, %1, %c2_i32 : i32
    cf.cond_br %2, ^bb2(%1, %0 : i32, i32), ^bb6
  ^bb2(%3: i32, %4: i32):  // 2 preds: ^bb1, ^bb4
    %5 = arith.addi %3, %c1_i32 : i32
    cf.br ^bb3
  ^bb3:  // pred: ^bb2
    %6 = arith.cmpi ult, %5, %c3_i32 : i32
    cf.cond_br %6, ^bb4, ^bb5
  ^bb4:  // pred: ^bb3
    %7 = arith.addi %4, %c1_i32 : i32
    cf.br ^bb2(%5, %7 : i32, i32)
  ^bb5:  // pred: ^bb3
    %8 = arith.addi %1, %c1_i32 : i32
    cf.br ^bb1(%4, %8 : i32, i32)
  ^bb6:  // pred: ^bb1
    return %0 : i32
  }
}
```

Note that after `^bb2`, this IR bears little resemblance to the input `test.mlir`, so it's difficult to determine whether or not `^bb2` and `^bb5` were merged with any `CHECK` rules.

In fact, here are the differences between bad (`main`) and good (this PR) when running `mlir-opt --canonicalize='enable-patterns=AnyPattern region-simplify=aggressive' test.mlir`:

image

If I were to write `CHECK` rules to look for the patterns on the right and `CHECK-NOT` the patterns on the left, it would be very difficult to maintain those going forward.

In summary, the other downstream IR changes from the pass make it very difficult to make this into a good and maintainable unit test in my opinion.

So I would like to propose that rather than making a unit test, I update the integration test with the simpler version of IR (i.e. `test.mlir` in this thread). Does that sound reasonable?

Comment (bmhowe23) [reply mlir/test/Integration/Dialect/ControlFlow/test-region-simplification.mlir]: OK, I replaced the integration test with a unit test here: 5b3ee9ffc8e5e5a8d8113316d328a2d6b8eb01d5. Please let me know if that's what you had in mind. (Thanks for your offline help, @joker-eph.)

**Comment by bmhowe23** [comment]
I believe I have addressed the only comment so far (moving an integration test into a unit test), and it's been ~2 weeks since there has been any discussion on this, so I figured I'd ping for review again.

File: mlir/test/Transforms/test-simplify-without-canonicalize.mlir [hunk 2]
```diff
@@ -0,0 +1,33 @@
+// RUN: mlir-opt --canonicalize='enable-patterns=AnyPattern region-simplify=aggressive' %s | FileCheck %s
+
+// Perform merge checks without performing canonicalization prior to simplification
+
+// This test should not merge ^bb2 and ^bb5, despite the fact that they are
+// identical because %4 is used outside of ^bb2.
```
Comment (joker-eph) [inline mlir/test/Transforms/test-simplify-without-canonicalize.mlir:6]: So on this example, there is no different in block merging before and after the patch is there? Like the same number of blocks are present in the output as far as I can tell?

Comment (bmhowe23) [reply mlir/test/Transforms/test-simplify-without-canonicalize.mlir:6]: The number of blocks is the same before and after the patch, but the `CHECK` item fails without the patch.

Comment (joker-eph) [reply mlir/test/Transforms/test-simplify-without-canonicalize.mlir:6]: Right, my point is that the comment is misleading here.

What about this reduction of the test case:

```
module {
  func.func @nested_loop(%arg0: i32, %arg1: i32, %arg2: i32, %arg3: i32, %arg4: i32, %arg5: i1) {
  // Irreducible control-flow: enter the middle of the loop in LoopBody_entry here.
    "test.foo_br"(%arg0, %arg4)[^LoopBody_entry] : (i32, i32) -> ()

  // Loop exit condition: jump to exit or 
  ^Loop_header:  // 2 preds: ^bb2, ^bb3
    // Consumes the block arg from LoopBody_entry
    // Because of this use here, we can't merge the two blocks below.
    "test.foo_br2"(%0)[^EXIT, ^LoopBody_entry, ^LoopBody_other] : (i32) -> ()

  // LoopBody_entry is jumped in from the entry block (bb0) and Loop_header
  // It **dominates** the Loop_header.
  ^LoopBody_entry(%0: i32):  // 2 preds: ^bb0, ^Loop_header
   // CHECK: test.foo
    %1 = "test.foo"(%0) : (i32) -> i32
    cf.br ^Loop_header

  // Other block inside the loop, not dominating the header
  ^LoopBody_other(%2: i32):  // pred: ^Loop_header
   // CHECK: test.foo
    %3 = "test.foo"(%2) : (i32) -> i32
    cf.br ^Loop_header

  ^EXIT:  // pred: ^Loop_header
    return
  }
}
```


Seems to me that it's more minimal and also more clear: we actually check that the block merge does not happen here.

Comment (bmhowe23) [reply mlir/test/Transforms/test-simplify-without-canonicalize.mlir]: Yes, that reduced test case looks reasonable to me, with the exception that - as written - it would pass without this PR because the `CHECK: test.foo` lines can accidentally match the `test.foo_br*` lines. If it's ok with you, I think we should change `test.foo` to `test.bar` so they don't accidentally match the wrong operations.

Updated in 395aeacf9ab973856279af5b0231f86f75c3cc2a.

**Review by joker-eph — APPROVED** [review]
LG, thanks

**Comment by bmhowe23** [comment]
Thanks, @joker-eph. I don't have commit privileges in this repo. Would you mind pushing this for me? Please let me know if I should do any final "update branch" or rebase.
